src/Database/LSMTree/Internal/MergeSchedule.hs

```diff
@@ -839,21 +693,25 @@ addRunToLevels tr conf@TableConfig{..} resolve hfs hbio root uc r0 reg levels =
             ir' <- newMerge LevelLevelling Merge.LastLevel ln (rs' `V.snoc` r)
             pure $! Level ir' V.empty `V.cons` V.empty
 
-    expectCompletedMergeTraced :: LevelNo -> IncomingRun m h
-                               -> m (Ref (Run m h))
-    expectCompletedMergeTraced ln ir = do
-      r <- expectCompletedMerge reg ir
+    expectCompletedMerge :: LevelNo -> IncomingRun m h -> m (Ref (Run m h))
+    expectCompletedMerge ln ir = do
+      r <- case ir of
+        Single r   -> pure r
+        Merging mr -> do
+          r <- allocateTemp reg (MR.expectCompleted mr) releaseRef
+          freeTemp reg (releaseRef mr)
+          pure r
       traceWith tr $ AtLevel ln $
         TraceExpectCompletedMerge (Run.runFsPathsNumber r)
       pure r
 
-    -- TODO: refactor, pull to top level?
+    -- Takes ownership of the runs passed.
```

Comment:
Might as well remove this comment, since it's not going to be true once we fix the TODO below

Reply:
IMO it makes the commit that fixes it clearer (we see in the diff what the old behaviour was). If it ends up squashed it doesn't matter either way.

Reply:
Ah, looking at this again, `newMerge` here actually consumes the references. Otherwise we'd have to do the `releaseTemp` at each call site instead, 6 times. Either way, it's up to date with the implementation.